ControllerUnpublishVolume should not fail when the corresponding vm is already deleted

chethanv28 opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR is addressing the issue where ControllerUnpublishVolume should not fail when the corresponding vm is already deleted. This is handled by handling the node not found error in ControllerUnpublishVolume

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #526

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
ControllerUnpublishVolume should not fail when the corresponding vm is already deleted

```

SandeepPissay commented:
@chethanv28 the code changes looks good to me. Can you add details on what testing was done to test the code change works as expected and  fixes the issue?

/approve

chethanv28 commented:
jtest block-vanilla

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 34 of 155 Specs in 5554.149 seconds
FAIL! -- 33 Passed | 1 Failed | 0 Pending | 121 Skipped
--- FAIL: TestE2E (5554.18s)
FAIL

Ginkgo ran 1 suite in 1h32m57.299660723s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/236/vsphere-csi-driver`

```

chethanv28 commented:
> @chethanv28 the code changes looks good to me. Can you add details on what testing was done to test the code change works as expected and fixes the issue?
> 
> /approve

Ran e2e tests. See: https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/561#issuecomment-755085081
```
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
Ran 34 of 155 Specs in 5554.149 seconds
FAIL! -- 33 Passed | 1 Failed | 0 Pending | 121 Skipped
--- FAIL: TestE2E (5554.18s)
FAIL

Ginkgo ran 1 suite in 1h32m57.299660723s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-block-vanilla/Results/236/vsphere-csi-driver`
```

The one failure is related to e2e test having an incorrect check.
```
Volume Expansion Test
/home/worker/workspace/github-csi-block-vanilla/Results/236/vsphere-csi-driver/tests/e2e/vsphere_volume_expansion.go:45
  [csi-block-vanilla] [csi-guest] Verify online volume expansion not allowed [It]
  /home/worker/workspace/github-csi-block-vanilla/Results/236/vsphere-csi-driver/tests/e2e/vsphere_volume_expansion.go:135

  Expected an error to have occurred.  Got:
      <nil>: nil
```

The one test failure is not related to this PR
cc: @SandeepPissay

divyenpatel commented on pkg/csi/service/vanilla/controller.go:
```diff
@@ -787,6 +788,10 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 	node, err := c.nodeMgr.GetNodeByName(ctx, req.NodeId)
 	if err != nil {
+		// ControllerUnpublishVolume should succeed when the corresponding vm is already deleted.
+		if err == cnsnode.ErrNodeNotFound {
```
We are returning `ErrNodeNotFound` only when we are unable to load node name to UUID mapping in the GetNodeByName() function. 

I think this will not help, when someone power down node VM, and delete it from the VM inventory without deleting node resource from the kubernetes cluster.

and if someone just deletes Kubernetes Node object without bringing down node VMs, we will end up in the situation, that we will not be able to create a new pod using the same volume on another node, as the volume is physically attached to some other VM on the vCenter.

Can you update PR with testing detail for the above cases?

SandeepPissay replied:
>I think this will not help, when someone power down node VM, and delete it from the VM inventory without deleting node resource from the kubernetes cluster.

Shouldn't Kubernetes delete the node object in this case after talking to CPI? Lets test and see whether this is handled correctly.

> and if someone just deletes Kubernetes Node object without bringing down node VMs
What is the use case for this? You mean the k8s admin accidentally deletes the wrong node in Kubernetes?

divyenpatel replied:
> Shouldn't Kubernetes delete the node object in this case after talking to CPI? Lets test and see whether this is handled correctly.

Kubernetes should delete the node object, after CCM returns `false with nil error` from `InstanceExistsByProviderID`
https://github.com/kubernetes/cloud-provider-vsphere/blob/master/pkg/cloudprovider/vsphere/instances.go#L157-L181

and once node object is deleted we have informer to delete mapping from local cache map. need to check what interval `InstanceExistsByProviderID` is being called by Kubernetes.


> What is the use case for this? You mean the k8s admin accidentally deletes the wrong node in Kubernetes?

Yes accidental deletion of k8s node object by admin.

chethanv28 replied:
> that we will not be able to create a new pod using the same volume on another node, as the volume is physically attached to some other VM on the vCenter.

Can confirm that a new pod can be created successfully using the same volume on another node. Added findings below:
https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/561#issuecomment-769993798

divyenpatel reviewed (changes requested):
Can you please address https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/561#discussion_r552894941

chethanv28 commented:
Created SC, PVC, POD
```
# kubectl create -f sc.yaml 
storageclass.storage.k8s.io/sc-0 created

root@k8s-master-234:~# kubectl create -f pvc.yaml 
persistentvolumeclaim/pvc-0 created

root@k8s-master-234:~# vi pod.yaml
root@k8s-master-234:~# kubectl create -f pod.yaml 
pod/pod-0 created

root@k8s-master-234:~# kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
pod-0   1/1     Running   0          2m13s

root@k8s-master-234:~# kubectl describe pods
Name:         pod-0
Node:         k8s-node-0397/10.186.74.138
Volumes:
  test-volume:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  pvc-0
    
  Normal  Started                 2m     kubelet, k8s-node-0397   Started container test-container
```

Get the node info to delete the node vm from VC
```
root@k8s-master-234:~# kubectl get nodes -o wide
NAME             STATUS   ROLES    AGE   VERSION   INTERNAL-IP     EXTERNAL-IP     OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
k8s-master-234   Ready    master   22d   v1.19.0   10.186.65.33    10.186.65.33    Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
k8s-node-0397    Ready    <none>   22d   v1.19.0   10.186.74.138   10.186.74.138   Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
k8s-node-0491    Ready    <none>   22d   v1.19.0   10.186.79.19    10.186.79.19    Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
k8s-node-06      Ready    <none>   22d   v1.19.0   10.186.78.61    10.186.78.61    Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8


root@k8s-master-234:~# kubectl get volumeattachments 
NAME                                                                   ATTACHER                 PV                                         NODE            ATTACHED   AGE
csi-964e249844d02a5ed1e60866691a5138555390bad4ab65cc6e0b879172e73644   csi.vsphere.vmware.com   pvc-f6115811-178e-4926-868a-5b437823d8a2   k8s-node-0397   true       3m20s


root@k8s-master-234:~# kubectl describe volumeattachments csi-964e249844d02a5ed1e60866691a5138555390bad4ab65cc6e0b879172e73644
Name:         csi-964e249844d02a5ed1e60866691a5138555390bad4ab65cc6e0b879172e73644
Spec:
  Attacher:   csi.vsphere.vmware.com
  Node Name:  k8s-node-0397                 <<<<<<<<<<<=====================
  Source:
    Persistent Volume Name:  pvc-f6115811-178e-4926-868a-5b437823d8a2
Status:
  Attached:  true
  Attachment Metadata:
    Disk UUID:  6000c29e431c5612fc68092cb1db8d84
    Type:       vSphere CNS Block Volume
Events:         <none>
```

Deleted k8s-node-0397 - vm from VC:
```
root@k8s-master-234:~# kubectl get nodes -o wide
NAME             STATUS   ROLES    AGE   VERSION   INTERNAL-IP    EXTERNAL-IP    OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
k8s-master-234   Ready    master   22d   v1.19.0   10.186.65.33   10.186.65.33   Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
k8s-node-0491    Ready    <none>   22d   v1.19.0   10.186.79.19   10.186.79.19   Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
k8s-node-06      Ready    <none>   22d   v1.19.0   10.186.78.61   10.186.78.61   Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
```

Pod gets deleted automatically
```
root@k8s-master-234:~# kubectl get pods
No resources found in default namespace.
```

Deleted the dangling VA:
```
kubectl delete volumeattachments csi-964e249844d02a5ed1e60866691a5138555390bad4ab65cc6e0b879172e73644
volumeattachment.storage.k8s.io "csi-964e249844d02a5ed1e60866691a5138555390bad4ab65cc6e0b879172e73644" deleted
```

Observed the existing ControllerUnPublish failure:
```
{"level":"info","time":"2021-01-29T18:27:52.210166319Z","caller":"vanilla/controller.go:732","msg":"ControllerUnpublishVolume: called with args {VolumeId:033ee9fb-13b2-4b09-a535-c3a4d408917a NodeId:k8s-node-0397 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}","TraceId":"c58a1ee7-1544-4018-bd36-8c91e60c62aa"}
{"level":"error","time":"2021-01-29T18:27:52.423552116Z","caller":"node/manager.go:133","msg":"Node not found with nodeName k8s-node-0397","TraceId":"c58a1ee7-1544-4018-bd36-8c91e60c62aa","stacktrace":"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/node.(*defaultManager).GetNodeByName\n\t/build/pkg/common/cns-lib/node/manager.go:133\nsigs.k8s.io/vsphere-csi-driver/pkg/csi/service/vanilla.(*Nodes).GetNodeByName\n\t/build/pkg/csi/service/vanilla/nodes.go:107\nsigs.k8s.io/vsphere-csi-driver/pkg/csi/service/vanilla.(*controller).ControllerUnpublishVolume\n\t/build/pkg/csi/service/vanilla/controller.go:788\ngithub.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler.func1\n\t/go/pkg/mod/github.com/container-storage-interface/spec@v1.2.0/lib/go/csi/csi.pb.go:5200\ngithub.com/rexray/gocsi/middleware/serialvolume.(*interceptor).controllerUnpublishVolume\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/middleware/serialvolume/serial_volume_locker.go:141\ngithub.com/rexray/gocsi/middleware/serialvolume.(*interceptor).handle\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/middleware/serialvolume/serial_volume_locker.go:88\ngithub.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/utils/utils_middleware.go:99\ngithub.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handleServer.func1\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/middleware/specvalidator/spec_validator.go:178\ngithub.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handle\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/middleware/specvalidator/spec_validator.go:218\ngithub.com/rexray/gocsi/middleware/specvalidator.(*interceptor).handleServer\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/middleware/specvalidator/spec_validator.go:177\ngithub.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/utils/utils_middleware.go:99\ngithub.com/rexray/gocsi.(*StoragePlugin).injectContext\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/middleware.go:231\ngithub.com/rexray/gocsi/utils.ChainUnaryServer.func2.1.1\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/utils/utils_middleware.go:99\ngithub.com/rexray/gocsi/utils.ChainUnaryServer.func2\n\t/go/pkg/mod/github.com/rexray/gocsi@v1.2.1/utils/utils_middleware.go:106\ngithub.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler\n\t/go/pkg/mod/github.com/container-storage-interface/spec@v1.2.0/lib/go/csi/csi.pb.go:5202\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.1\n\t/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722"}

```

Deployed new driver with images built from this PR. Observed the dangling VA is deleted this time:

```
root@k8s-master-234:~# kubectl get volumeattachments 
No resources found
```

Deployed a new pod with the same volume:
```
root@k8s-master-234:~# kubectl create -f pod.yaml 
pod/pod-0 created

kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
pod-0   1/1     Running   0          71s
root@k8s-master-234:~# kubectl get volumeattachments 
NAME                                                                   ATTACHER                 PV                                         NODE            ATTACHED   AGE
csi-119f9e265d326d0da17ca64cc73f736cd363f328dbb8cde3c196e3972b7e9dde   csi.vsphere.vmware.com   pvc-f6115811-178e-4926-868a-5b437823d8a2   k8s-node-0491   true       76s

kubectl describe volumeattachments csi-119f9e265d326d0da17ca64cc73f736cd363f328dbb8cde3c196e3972b7e9dde
Name:         csi-119f9e265d326d0da17ca64cc73f736cd363f328dbb8cde3c196e3972b7e9dde
Namespace:    
Labels:       <none>
Annotations:  csi.alpha.kubernetes.io/node-id: k8s-node-0491   <<<<<<===========================
```

Deleted the node vm - k8s-node-0491
```
root@k8s-master-234:~# kubectl get nodes -o wide
NAME             STATUS   ROLES    AGE   VERSION   INTERNAL-IP    EXTERNAL-IP    OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
k8s-master-234   Ready    master   22d   v1.19.0   10.186.65.33   10.186.65.33   Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
k8s-node-06      Ready    <none>   22d   v1.19.0   10.186.78.61   10.186.78.61   Ubuntu 20.04.1 LTS   5.4.0-42-generic   docker://19.3.8
```

Pod gets deleted automatically
```
root@k8s-master-234:~# kubectl get pods
No resources found in default namespace.
```

Deleted the dangling VA
```
kubectl delete volumeattachments csi-119f9e265d326d0da17ca64cc73f736cd363f328dbb8cde3c196e3972b7e9dde 
```

This time the VA does not exist. It gets deleted since ControllerUnPublish succeeded
```
root@k8s-master-234:~# kubectl get volumeattachments 
No resources found
```


cc: @SandeepPissay @divyenpatel  Please find the testing details above^^

SandeepPissay commented on pkg/csi/service/vanilla/controller.go:
```diff
@@ -787,6 +788,11 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 	// Block Volume
 	node, err := c.nodeMgr.GetNodeByName(ctx, req.NodeId)
 	if err != nil {
+		// ControllerUnpublishVolume should succeed when the corresponding vm is already deleted.
```
I have a concern. There are cases where the node manager has not yet synced the nodes(maybe the node events got delayed) and the ControllerUnpublish is invoked. In this case, we incorrect return success that the detach is successful.

chethanv28 replied:
Observed both the times node manager update was immediate, after deleting vm from VC.

```
{"level":"info","time":"2021-01-29T18:24:33.9933385Z","caller":"node/manager.go:266","msg":"Successfully unregistered node with nodeName k8s-node-0397","TraceId":"1883f8f1-54ef-4722-bb45-81fc7821c67a"}
```

```
{"level":"info","time":"2021-01-29T18:59:00.966511893Z","caller":"node/manager.go:266","msg":"Successfully unregistered node with nodeName k8s-node-0491","TraceId":"090f7b3a-7e59-48c1-a169-45b134fb1d55"}
```

Also, we are querying the API server in GetNodeByName() if we anticipate any delay in the node events:
https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/common/cns-lib/node/manager.go#L139

SandeepPissay replied:
Here's the scenario. A pod is deleted when the CSI node is down. Now the CSI pod comes up, we now have a race between ControllerUnpublishVolume execution and node manager getting events for the current nodes in k8s. If ControllerUnpublishVolume runs first(before node manager is aware of any node), then they all succeed leaving behind the volumes attached to the node. The delete volume would never succeed since the volume is still attached. Do you agree?

chethanv28 replied:
Handled this by invoking get nodes on k8s in node manager

SandeepPissay commented:
> Pod gets deleted automatically

Shouldn't it get rescheduled? What is the restartPolicy for the Pod?

chethanv28 commented:
jtest block-vanilla

svcbot-qecnsdp commented:
Started Vanilla block pipeline...

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: testbed-deploy 

```

chethanv28 commented:
The code changes in this PR 529. Closing this PR since the vSphere CSI dev team anticipate that there could be more issues if we simply return successful response for ControllerUnpublishVolume when nodes are deleted.
